shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/TemporaryTerminalRewriter.kt

```diff
@@ -0,0 +1,254 @@
+     *    non-typical pattern hits GL early-morning cross-branch trips)
+     * 4. but the predictions all have typical patterns instead of non-typical patterns (meaning
+     *    this hasn't been fixed upstream yet)
+     */
+    fun appliesToRoute(route: Route): Boolean {
+        val routeId = route.id
+        val isSubway = route.type.isSubway()
+
+        val routeHasAlert =
+            alerts.any { alert ->
+                alert.effect in setOf(Alert.Effect.Suspension, Alert.Effect.Shuttle) &&
+                    alert.anyInformedEntity { it.appliesTo(routeId = routeId) }
+            }
+
+        val routeSchedules = schedulesByRoute[routeId].orEmpty()
+        val scheduledPatterns = routeSchedules.mapNotNullTo(mutableSetOf()) { it.routePattern() }
+        val routePatterns = routePatternsByRoute[routeId].orEmpty()
+        val scheduleMissingTypical =
+            routePatterns
+                .filter { it.typicality == RoutePattern.Typicality.Typical }
+                .any { it !in scheduledPatterns }
+        val scheduleHasNontypical =
+            scheduledPatterns.any { it.typicality != RoutePattern.Typicality.Typical }
+        val scheduleReplacedTypical = scheduleMissingTypical && scheduleHasNontypical
+
+        val routePredictions = predictionsByRoute[routeId].orEmpty()
+        val predictionsAlwaysTypical =
+            routePredictions.isNotEmpty() &&
+                routePredictions.all {
+                    it.routePattern()?.typicality == RoutePattern.Typicality.Typical
+                }
+
+        return isSubway && routeHasAlert && scheduleReplacedTypical && predictionsAlwaysTypical
+    }
+
+    /**
+     * Checks if this route pattern is not typical, points in the same direction as [fullPattern],
+     * contains [stopId], and contains a subsequence of [fullPatternStopIds].
+     */
+    fun RoutePattern.isTruncationOf(
+        fullPattern: RoutePattern,
+        fullPatternStopIds: List<String>,
+        stopId: String
+    ): Boolean {
+        if (typicality == RoutePattern.Typicality.Typical) return false
+        if (directionId != fullPattern.directionId) return false
+        val truncatedPatternStopIds =
+            globalData.trips[representativeTripId]?.stopIds ?: return false
+        return truncatedPatternStopIds.contains(stopId) &&
+            fullPatternStopIds.containsSubsequence(truncatedPatternStopIds)
+    }
+
+    /**
+     * Finds a truncated pattern for [fullPattern] that includes [stopId]. Since at time of writing
+     * we only fetch scheduled trips for the remainder of the day, does not require that the
+     * truncated pattern actually appear in the schedule unless there are multiple plausible
+     * patterns (which could happen if the rating includes both a current disruption and a future
+     * disruption on the same line). If there is no plausible truncated pattern, or if there are
+     * several, does not truncate [fullPattern].
+     */
```

Comment:
This has taken me a few reads to grasp, the unit tests help. Perhaps the name `findTruncatedPattern` or `matchTruncatedPattern` for this and `truncatePatternsAtStop` would be clearer that this isn't doing any truncation itself

Reply:
Renamed this function.